lib/active_merchant/billing/credit_card_methods.rb

```diff
@@ -431,19 +437,32 @@ def valid_luhn?(numbers) #:nodoc:
           sum = 0
 
           odd = true
-          numbers.reverse.bytes.each do |number|
+          numbers.reverse.bytes.each do |bytes|
             if odd
               odd = false
-              sum += ODD_LUHN_VALUE[number]
+              sum += BYTES_TO_DIGITS[bytes]
             else
               odd = true
-              sum += EVEN_LUHN_VALUE[number]
+              sum += BYTES_TO_DIGITS_DOUBLED[bytes]
             end
           end
 
           sum % 10 == 0
         end
 
+        def valid_luhn_with_check_digit?(numbers, check_digit)
+          sum = 0
+
+          doubler = true
+
+          numbers.reverse.bytes.each do |bytes|
+            doubler ? sum += BYTES_TO_DIGITS_DOUBLED[bytes] : sum += BYTES_TO_DIGITS[bytes]
+            doubler = !doubler
+          end
+
+          10 - (sum % 10) == check_digit.to_i
+        end
+
```

Comment:
Alternatively, we could rewrite `valid_luhn?` to switch based on whether we are passing a check digit or not. That would look something like:

```
def valid_luhn?(numbers, check_digit = nil)
  sum = 0

  doubler = check_digit.present?

  numbers.reverse.bytes.each do |bytes|
    doubler ? sum += BYTES_TO_DIGITS_DOUBLED[bytes] : sum += BYTES_TO_DIGITS[bytes]
    doubler = !doubler
  end

  if check_digit.present?
    10 - (sum % 10) == check_digit.to_i
  else
    sum % 10 == 0
  end
end
```

I didn't do this because it's probably more prudent to avoid refactoring the `valid_luhn?` method which almost every card number hits, and it doesn't necessarily make things more clear. But I'm open to differing opinions on this.

Reply:
I could go either way on this. I think the refactor example you gave would be easy enough to do and we have the tests coverage to verify implementation, so I think the risk is relatively low.

Reply:
I decided to leave it as-is out of an abundance of caution.